File: Woche1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void) {
    long ccard;
    long divisor = 10;

    do {
        ccard = get_long_long("Credit: ");
    } while (ccard <= 0);
    
    long workingCC = ccard;
    int sum = 0;
    int count = 0;

    while (workingCC > 0) {
        int lastDigit = workingCC % 10;
        sum = sum + lastDigit; 
        workingCC = workingCC / 100;
    }

    workingCC = ccard / 10;
    while (workingCC > 0) {
        int lastDigit = workingCC % 10;
        int timesTwo = lastDigit * 2;
        sum = sum + (timesTwo % 10) + (timesTwo / 10);
        workingCC = workingCC / 100;
    }

    workingCC = ccard;
    while (workingCC != 0) {
        workingCC = workingCC / 10;
        count++;
    }
    
    for (int i = 0; i < count - 2; i++) {
        divisor = divisor * 10;
    }

    int firstDigit = ccard / divisor;
    int firstTwoDigits = ccard / (divisor / 10); 

    if (sum % 10 == 0) {

        bool CardIsMasterCard = count == 16 && (firstTwoDigits == 51 || firstTwoDigits == 52 || firstTwoDigits == 53 || firstTwoDigits == 54 || firstTwoDigits == 55);
        if(CardIsMasterCard) {
            printf("MasterCard\n");
            return 0;
        }
        
        bool CardIsAmericanExpress = (firstTwoDigits == 34 || firstTwoDigits == 37) && count == 15;
        if(CardIsAmericanExpress) {
            printf("AmericanExpress\n");
            return 0;
        }
        
        bool CardIsVisa = (count == 13 || count == 16) && firstDigit == 4;
        if(CardIsVisa) {
            printf("Visa\n");
            return 0;
        }

        printf("Invalid\n");

    } else {
        printf("Invalid\n");
    }
}

File: Woche6/sentimental-credit/credit.py
from cs50 import get_int


def input_number(message):
    while True:
        try:
            user_input = get_int(message)

            if user_input <= 1:
                print("Invalid")
                return

        except ValueError:
            continue
        else:
            return user_input


def check_which_type_card_is():
    if separate_digits % 10 == 0:
        card_is_master_card = count == 16 and (
                first_two_digits == 51
                or first_two_digits == 52
                or first_two_digits == 53
                or first_two_digits == 54
                or first_two_digits == 55
        )

        if card_is_master_card:
            print("MasterCard")
            return

        card_is_american_express = (first_two_digits == 34 or first_two_digits == 37) and count == 15
        if card_is_american_express:
            print("AmericanExpress")
            return

        card_is_visa = (count == 13 or 16) and first_digit == 4
        if card_is_visa:
            print("Visa")
            return

        print("Invalid")

    else:
        print("Invalid 2")


separate_digits = 0
count = 0
divisor = 10


credit_card_number = input_number("Credit Number: ")

workingCC = credit_card_number

while workingCC > 0:
    last_digit = workingCC % 10
    separate_digits = separate_digits + last_digit
    workingCC = workingCC // 100


workingCC = credit_card_number // divisor
while workingCC > 0:
    last_digit = workingCC % 10
    times_two = last_digit * 2
    separate_digits = separate_digits + (times_two % 10) + (times_two // 10)
    workingCC = workingCC // 100

workingCC = credit_card_number
while workingCC != 0:
    workingCC = workingCC // 10
    count += 1


i = 0
while i < count - 2:
    divisor = divisor * 10
    i += 1


first_digit = credit_card_number // divisor
first_two_digits = credit_card_number // (divisor // 10)


check_which_type_card_is()
